### PR TITLE
WHFHRI-671

### PR DESCRIPTION
**Pull request by alisakotliarova**
## PR Requirements:
 - [ ] The code must be reviewed by at least one Maintainer
 - [ ] JIRA Ticket: https://jira.wh-sdlc.watson-health.ibm.com/browse/JIRA_TICKET_HERE
 - [ ] 90 % test code coverage
 - [ ] README updates

### CODE REVIEW

**Comment by aramcodz** [comment]
Two Reminders:
1. We are still blocked on the Static Code Analysis implementation (that uses sonarcloud-github-action); see --> https://github.com/Alvearie/hri-mgmt-api/blob/develop/.github/workflows/code-scans.yml
2. Alisa is currently blocked on the required Open Source/Vulnerability scan until I complete https://jira.wh-sdlc.watson-health.ibm.com/browse/WHFHRI-699

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+      - uses: actions/checkout@v2
+
+      - name: Set up adoptopenjdk8
+        uses: actions/setup-java@v2
+        with:
+          java-version: '8'
+          distribution: 'adopt'
+          cache: 'gradle'
+
```
Comment (aramcodz) [inline .github/workflows/release.yml:29]: Does this action code correctly set the tags for releases?  I feel like perhaps we are missing some operation here.  For example, I think we need something to set the `ACTIONS_TAG` Env Var.

File: .github/workflows/ci-workflow.yml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  push:
+    branches: [ '**' ]
+  workflow_dispatch:
+
+concurrency: ci-${{ github.ref }}
+
+jobs:
+  buildAndTest:
+    name: Build, Test, and Package
+    runs-on: ubuntu-latest
```
Comment (aramcodz) [inline .github/workflows/ci-workflow.yml:13]: @David-N-Perkins: other than the dependency vulnerabiliuty scan, is there any other operation we are missing from this `ci-workflow`?

File: build.gradle [hunk 3]
```diff
@@ -55,15 +54,26 @@ tasks.withType(JavaCompile) {
...
+    }
+}
 
-            url "TBD/$repo/"
+repositories {
+    maven {
+        url = uri("https://maven.pkg.github.com/Alvearie/hri-api-spec")
+        credentials {
+            username = findProperty("GITHUB_ACTOR") ?: System.getenv("GITHUB_ACTOR")
+            password = findProperty("GITHUB_TOKEN") ?: System.getenv("GITHUB_TOKEN")
```
Comment (aramcodz) [inline build.gradle:76]: It looks like we now have 2 different `repositories` sections in this file, one abive at line 58 and this one?  

Do we need both? Is this one redundant?

File: build.gradle [hunk 4]
```diff
@@ -92,7 +102,7 @@ dependencies {
     // Dependencies that library users should include in their job
     // shadow jar
     // --------------------------------------------------------------
-    implementation "org.alvearie.hri:hri-api-batch-notification:develop-SNAPSHOT"
+    implementation "org.alvearie.hri:hri-api-batch-notification:WHFHRI-693-SNAPSHOT"
```
Comment (aramcodz) [inline build.gradle:105]: I think you do NOT want to change this dependency here. you want it to pull in `develop-SNAPSHOT`.

I think using `develop-SNAPSHOT` should work the same as using `WHFHRI-693-SNAPSHOT`.

File: build.gradle [hunk 5]
```diff
@@ -169,4 +182,3 @@ def getWorkingBranch() {
     println "Working branch: " + workingBranch
```
Comment (aramcodz) [inline build.gradle:182]: NOTE: Should probably remove line 182 before merging.

**Review by aramcodz — COMMENTED** [review]
A few questions/comments.

File: .github/workflows/ci-workflow.yml [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+      - name: Gradle build
+        run: |
+          chmod +x ./gradlew
+          ./gradlew clean build
+
+      - name: Gradle publish
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: ./gradlew publish --stacktrace
```
Comment (David-N-Perkins) [inline .github/workflows/ci-workflow.yml:38]: Is the `--stacktrace` flag needed?

File: .github/workflows/ci-workflow.yml [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+      - name: Set up adoptopenjdk8
+        uses: actions/setup-java@v2
+        with:
+          java-version: '8'
+          distribution: 'adopt'
+          cache: 'gradle'
+
+      - name: Gradle build
+        run: |
+          chmod +x ./gradlew
```
Comment (David-N-Perkins) [inline .github/workflows/ci-workflow.yml:32]: You should set this as executable and commit it, and then remove this line.

File: .github/workflows/ci-workflow.yml (hunk 2 above)
Comment (David-N-Perkins) [reply .github/workflows/ci-workflow.yml:13]: The vulnerability scan would go in different workflows. Sonar Cloud analysis could go in this workflow, but I put it in a different one for hri-mgmt-api.

File: .github/workflows/release.yml [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+      - name: Gradle build
+        run: |
+          chmod +x ./gradlew
+          ./gradlew clean build
+
+      - name: Gradle publish
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: ./gradlew publish --stacktrace
```
Comment (David-N-Perkins) [inline .github/workflows/release.yml:38]: Maybe I'm missing something, but this workflow looks the same as `ci-workflow.yml`.  Could we combine them? I think the key would be correctly determining the artifact ID in the Gradle build.  Branches would need the `-SNAPSHOT` suffix, while tags are supposed to match the version.

File: .github/workflows/release.yml (hunk 1 above)
Comment (David-N-Perkins) [reply .github/workflows/release.yml:29]: I agree. Either the tag needs to be extracted here, or the build.gradle logic needs to be updated.

File: build.gradle [hunk 9]
```diff
@@ -111,8 +121,8 @@ dependencies {
     testImplementation "org.apache.flink:flink-streaming-java_${scalaBinaryVersion}:${flinkVersion}:tests"
 
     testRuntimeOnly "com.vladsch.flexmark:flexmark-all:0.35.10"
-    testRuntime 'org.pegdown:pegdown:1.4.2'
-    testCompile "info.picocli:picocli:4.2.0"
+    testRuntimeOnly 'org.pegdown:pegdown:1.6.0'
+    testImplementation 'info.picocli:picocli:4.6.1'
```
Comment (David-N-Perkins) [inline build.gradle:125]: Why did you need to change these scopes? Are those the old dependency categories?

Comment (alisakotliarova) [reply build.gradle:131]: Yes, per the [Java library plugin documentation from Gradle](https://docs.gradle.org/3.4/userguide/java_library_plugin.html?_ga=2.49116771.473058149.1632775631-1137966675.1632157466#sec:java_library_configurations_graph), 

> The compile, testCompile, runtime and testRuntime configurations inherited from the Java plugin are still available but are deprecated. You should avoid using them, as they are only kept for backwards compatibility.

`testRuntimeOnly` is supposed to replace `testRuntime`, and `testImplementation` is supposed to replace `testCompile`.  If you try build with the old category names, it should show a warning that also says this.

File: .github/workflows/release.yml (hunk 8 above)
Comment (alisakotliarova) [reply .github/workflows/release.yml:37]: I modeled the two workflows after the ones in `hri-mgmt-api`.  Those are also similar but they have different vulnerability checks and scans.  I assumed that when Aram would add in those checks and scans to this branch that the functionality would diverge and they would be more different in the way the two are in `hri-mgmt-api`.  I can combine them instead.

**Review by FRicci618 — APPROVED** [review]

File: .github/workflows/release.yml (hunk 1 above)
Comment (alisakotliarova) [reply .github/workflows/release.yml]: I added the tag extraction ✅

File: build.gradle (hunk 3 above)
Comment (alisakotliarova) [reply build.gradle]: One of the sections I have found to be unnecessary, I have removed that section.

File: build.gradle (hunk 4 above)
Comment (alisakotliarova) [reply build.gradle]: I changed it again while troubleshooting again, it is now reverted ✅

File: .github/workflows/ci-workflow.yml (hunk 7 above)
Comment (alisakotliarova) [reply .github/workflows/ci-workflow.yml]: Done ✅

File: build.gradle (hunk 5 above)
Comment (alisakotliarova) [reply build.gradle]: Removed ✅

File: .github/workflows/ci-workflow.yml (hunk 6 above)
Comment (alisakotliarova) [reply .github/workflows/ci-workflow.yml]: Not when everything is working.  I have removed it. ✅

File: .github/workflows/release.yml (hunk 8 above)
Comment (aramcodz) [reply .github/workflows/release.yml]: > I modeled the two workflows after the ones in `hri-mgmt-api`. Those are also similar but they have different vulnerability checks and scans. I assumed that when Aram would add in those checks and scans to this branch that the functionality would diverge and they would be more different in the way the two are in `hri-mgmt-api`. I can combine them instead.

Yeah, @alisakotliarova makes a good point here. The OpenSource/Vulnerability scan will probably only happen in another (2nd) `pull_request.yml` workflow (as per this comment --> https://github.com/Alvearie/hri-flink-pipeline-core/pull/3/files#r716981860). 

So we should have a separate `pull_request.yml` workflow I modeled on https://github.com/Alvearie/hri-mgmt-api/blob/develop/.github/workflows/pull_request.yml, @David-N-Perkins. And, then the Vulnerability scan call will get put into that one.

File: build.gradle [hunk 10]
```diff
@@ -121,20 +122,11 @@ publishing {
...
             artifact jarTests
         }
     }
-
-    repositories {
-        maven {
-            url 'TBD'
-            credentials {
-                username findProperty('user') ?: System.getenv('user')
-                password findProperty('password') ?: System.getenv('password')
```
Comment (David-N-Perkins) [inline build.gradle:134]: I thought this needed to be updated with the GitHub repo for publishing to GitHub packages? Is that not the case?

Comment (aramcodz) [reply build.gradle:134]: > I thought this needed to be updated with the GitHub repo for publishing to GitHub packages? Is that not the case?

@alisakotliarova, I believe that is correct.

It should be something like:

```
               username = findProperty("GITHUB_ACTOR") ?: System.getenv("GITHUB_ACTOR")
               password = findProperty("GITHUB_TOKEN") ?: System.getenv("GITHUB_TOKEN")
```

See the validation/build.gradle here (on my PR-test branch) --> https://github.com/Alvearie/hri-flink-validation-fhir/tree/WHFHRI-699-test-pr/validation

Comment (aramcodz) [reply build.gradle:134]: > > I thought this needed to be updated with the GitHub repo for publishing to GitHub packages? Is that not the case?
> 
> @alisakotliarova, I believe that is correct.
> 
> It should be something like:
> 
> ```
>                username = findProperty("GITHUB_ACTOR") ?: System.getenv("GITHUB_ACTOR")
>                password = findProperty("GITHUB_TOKEN") ?: System.getenv("GITHUB_TOKEN")
> ```
> 
> See the validation/build.gradle here (on my PR-test branch) --> https://github.com/Alvearie/hri-flink-validation-fhir/tree/WHFHRI-699-test-pr/validation

THis has now been Fixed. I am marking Resolved.

**Comment by aramcodz** [comment]
> Two Reminders:
> 
> 1. We are still blocked on the Static Code Analysis implementation (that uses sonarcloud-github-action); see --> https://github.com/Alvearie/hri-mgmt-api/blob/develop/.github/workflows/code-scans.yml. We are waiting on Alvearie team support to get us access to the alvearie sonarcloud account for this Repo.
> 2. Alisa is currently blocked on the required Open Source/Vulnerability scan until I complete https://jira.wh-sdlc.watson-health.ibm.com/browse/WHFHRI-699

**_UPDATE:_** 
Item # 2 (from above) is now completed.
Item # 1 will be assigned as a task in Follow-on story

File: build.gradle (hunk 3 above)
Comment (alisakotliarova) [reply build.gradle]: The other section is actually needed for publishing, it has been added back

File: build.gradle (hunk 10 above)
Comment (aramcodz) [reply build.gradle]: Fixed now - I am marking resolved.

**Review by aramcodz — APPROVED** [review]

**Review by David-N-Perkins — COMMENTED** [review]
`pull_request.yml` needs to be updated with Arams example, which I think also requires some `build.gradle` updates.

**Pull request merged**